File: frontend/src/lib/atsScorer.js
/**
 * atsScorer.js — Client-side ATS scoring engine.
 *
 * scoreResume(text, jobDescription?) → ScoreResult
 * No API calls. All logic runs in the browser.
 */

import powerSkillsData from '../data/powerSkills.json';
import actionVerbsList from '../data/actionVerbs.json';
import sectionKeywords from '../data/sectionKeywords.json';

// ── Flatten skills for matching ──────────────────────────────────

function deepFlattenSkills(obj) {
  const result = [];
  for (const val of Object.values(obj)) {
    if (Array.isArray(val)) result.push(...val);
    else if (typeof val === 'object' && val !== null) result.push(...deepFlattenSkills(val));
  }
  return result;
}

const ALL_SKILLS = deepFlattenSkills(powerSkillsData).map(s => s.toLowerCase());
const SKILL_SET = new Set(ALL_SKILLS);
const ACTION_VERBS = new Set(actionVerbsList.map(v => v.toLowerCase()));
const WEAK_VERBS = new Set(['helped','worked','did','made','was','had','got','went','used','tried','handled','assisted','participated','responsible','involved']);

// ── Section detection ────────────────────────────────────────────

function detectSections(text) {
  const lines = text.split('\n');
  const sections = [];
  let current = { type: 'unknown', title: '', lines: [] };

  for (const line of lines) {
    const trimmed = line.trim();
    if (!trimmed) { current.lines.push(line); continue; }

    const lower = trimmed.toLowerCase().replace(/[^a-z\s]/g, '').trim();
    let matched = null;
    for (const [type, keywords] of Object.entries(sectionKeywords)) {
      if (keywords.some(kw => lower === kw || lower.startsWith(kw + ' ') || lower.endsWith(' ' + kw))) {
        matched = type;
        break;
      }
    }

    if (matched) {
      if (current.lines.length > 0 || current.type !== 'unknown') {
        sections.push({ ...current, content: current.lines.join('\n') });
      }
      current = { type: matched, title: trimmed, lines: [] };
    } else {
      current.lines.push(line);
    }
  }
  if (current.lines.length > 0 || current.type !== 'unknown') {
    sections.push({ ...current, content: current.lines.join('\n') });
  }
  return sections;
}

// ── 1. Keywords & Skills (30 pts) ────────────────────────────────

function matchSkill(skill, text, wordTokens) {
  if (skill.includes(' ')) {
    // Multi-word phrase: word boundary regex
    return getWordBoundaryRegex(skill).test(text);
  } else if (skill.length <= 2) {
    // Very short (R, C, AI, UX): exact token match only
    return wordTokens.has(skill);
  } else {
    // Normal: word boundary regex
    return getWordBoundaryRegex(skill).test(text);
  }
}

function scoreKeywords(text, jobDescription) {
  const lower = text.toLowerCase();
  const wordTokens = new Set(lower.split(/[\s,;|/()[\]{}<>:]+/).filter(Boolean));

  const matched = ALL_SKILLS.filter(s => matchSkill(s, lower, wordTokens));
  const matchedSet = [...new Set(matched)];

  let jdTerms = [];
  if (jobDescription) {
    const jdLower = jobDescription.toLowerCase();
    const jdTokens = new Set(jdLower.split(/[\s,;|/()[\]{}<>:]+/).filter(Boolean));
    jdTerms = ALL_SKILLS.filter(s => matchSkill(s, jdLower, jdTokens));
    jdTerms = [...new Set(jdTerms)];
  }

  const expected = Math.max(jdTerms.length || 12, 12);
  const missing = (jdTerms.length > 0 ? jdTerms : ALL_SKILLS.slice(0, 30))
    .filter(s => !matchedSet.includes(s)).slice(0, 15);

  const pts = Math.min(30, Math.round((matchedSet.length / expected) * 30));
  return {
    score: pts, max: 30,
    matched: matchedSet, missing,
    matchPct: Math.round((matchedSet.length / expected) * 100)
  };
}

// ── 2. Formatting & Parseability (20 pts) ────────────────────────

function scoreFormatting(text) {
  let pts = 20;
  const issues = [];

  if (/[│┃┆┇┊┋|]{2,}/m.test(text) || /\t.*\t.*\t/m.test(text)) {
    pts -= 5; issues.push('Table-like layout detected — ATS often misreads tables');
  }
  if (/\[image\]|\[graphic\]|\.png|\.jpg|\.svg/i.test(text)) {
    pts -= 5; issues.push('Image/graphic references detected — ATS cannot parse images');
  }
  const lines = text.split('\n');
  const wideLines = lines.filter(l => l.length > 120);
  if (wideLines.length > 3) {
    pts -= Math.min(4, wideLines.length); issues.push('Lines exceed 120 characters — may cause parsing issues');
  }
  const fancyBullets = (text.match(/[◆▶►★✦✧●○◉⬤⚫⬥◈▪▫]/g) || []).length;
  if (fancyBullets > 0) {
    pts -= Math.min(3, fancyBullets); issues.push('Fancy bullets detected — use standard dashes or dots');
  }

  const foundHeaders = Object.entries(sectionKeywords).filter(([, kws]) =>
    kws.some(kw => text.toLowerCase().includes(kw))
  );
  if (foundHeaders.length < 3) {
    pts -= 5; issues.push('Missing standard section headers');
  }

  return { score: Math.max(0, pts), max: 20, issues };
}

// ── 3. Contact Completeness (10 pts) ─────────────────────────────

function scoreContact(text) {
  let pts = 0;
  const details = {};

  if (/[\w.-]+@[\w.-]+\.\w{2,}/i.test(text)) { pts += 3; details.email = true; }
  if (/(\+?\d[\d\s\-().]{7,}\d)/.test(text)) { pts += 3; details.phone = true; }
  if (/linkedin\.com\/in\/[\w-]+/i.test(text)) { pts += 2; details.linkedin = true; }
  if (/\b[A-Z][a-z]+,?\s*[A-Z]{2}\b/.test(text) || /\b\d{5,6}\b/.test(text)) {
    pts += 2; details.location = true;
  }
  if (/github\.com\/[\w-]+/i.test(text)) { pts += 1; }

  return { score: Math.min(10, pts), max: 10, details };
}

// ── 4. Section Structure (15 pts) ────────────────────────────────

function scoreStructure(text) {
  const lower = text.toLowerCase();
  const required = ['summary', 'experience', 'education', 'skills'];
  const found = [];
  const missing = [];

  for (const sec of required) {
    const kws = sectionKeywords[sec] || [];
    if (kws.some(kw => lower.includes(kw))) found.push(sec);
    else missing.push(sec);
  }

  const pts = Math.round(found.length * 3.75);
  return { score: pts, max: 15, found, missing };
}

// ── 5. Quantified Achievements (15 pts) ──────────────────────────

function scoreAchievements(text) {
  const quantRegex = /\d+[%x×kmb$]|\$[\d,.]+|\d+\s*(?:users|clients|projects|customers|employees|members|teams|revenue|sales|leads)/gi;
  const matches = text.match(quantRegex) || [];
  const count = matches.length;

  let pts;
  if (count === 0) pts = 0;
  else if (count <= 2) pts = 6;
  else if (count <= 5) pts = 10;
  else if (count <= 9) pts = 13;
  else pts = 15;

  const lines = text.split('\n').filter(l => quantRegex.test(l)).slice(0, 5);
  return { score: pts, max: 15, count, examples: lines };
}

// ── 6. Action Verbs (5 pts) ──────────────────────────────────────

function scoreActionVerbs(text) {
  const lines = text.split('\n').map(l => l.trim()).filter(l =>
    /^[-•–—*]/.test(l) || /^\d+[.)]/.test(l) || (l.length > 20 && l.length < 200)
  );
  if (lines.length === 0) return { score: 0, max: 5, coverage: 0, weak: [] };

  let strong = 0, weakFound = [];
  for (const line of lines) {
    const first = line.replace(/^[-•–—*\d.)]+\s*/, '').split(/\s+/)[0]?.toLowerCase();
    if (!first) continue;
    if (ACTION_VERBS.has(first)) strong++;
    else if (WEAK_VERBS.has(first)) weakFound.push(first);
  }

  const coverage = lines.length > 0 ? (strong / lines.length) * 100 : 0;
  let pts;
  if (coverage >= 80) pts = 5;
  else if (coverage >= 60) pts = 3;
  else if (coverage >= 40) pts = 2;
  else pts = 0;

  return { score: pts, max: 5, coverage: Math.round(coverage), weak: [...new Set(weakFound)] };
}

// ── 7. Length & Density (5 pts) ──────────────────────────────────

function scoreLength(text) {
  const wordCount = text.split(/\s+/).filter(Boolean).length;
  let pts, message;

  if (wordCount >= 400 && wordCount <= 700) { pts = 5; message = 'Optimal length for a 1-page resume'; }
  else if (wordCount >= 701 && wordCount <= 900) { pts = 3; message = 'Slightly dense — consider trimming'; }
  else if (wordCount >= 901 && wordCount <= 1200) { pts = 5; message = 'Good length for a 2-page resume'; }
  else if (wordCount > 1200) { pts = 2; message = 'Too long — most ATS prefer 1–2 pages'; }
  else if (wordCount < 250) { pts = 2; message = 'Too short — add more detail'; }
  else { pts = 3; message = 'Acceptable length'; }

  return { score: pts, max: 5, wordCount, message };
}

// ── Suggestion generator ─────────────────────────────────────────

function generateSuggestions(categories) {
  const suggestions = [];
  let id = 0;

  const add = (priority, category, title, detail, sectionTarget = null, diff = null) => {
    suggestions.push({ id: `s${id++}`, priority, category, title, detail, sectionTarget, diff });
  };

  const { keywords, formatting, contact, structure, achievements, actionVerbs, length } = categories;

  // Keywords
  if (keywords.score < 15) {
    add('critical', 'keywords', 'Add more relevant skills', `Only ${keywords.matched.length} skills detected. Add skills like: ${keywords.missing.slice(0,5).join(', ')}.`, 'skills');
  } else if (keywords.score < 24) {
    add('important', 'keywords', 'Expand your skillset', `You're missing key terms: ${keywords.missing.slice(0,4).join(', ')}. Weave them into your experience bullets.`, 'skills');
  }
  if (keywords.missing.length > 0 && keywords.score >= 24) {
    add('nice', 'keywords', 'Consider adding niche skills', `Skills like ${keywords.missing.slice(0,3).join(', ')} could strengthen your profile.`, 'skills');
  }

  // Formatting
  if (formatting.issues.length > 0) {
    for (const issue of formatting.issues.slice(0, 3)) {
      const p = formatting.score < 10 ? 'critical' : 'important';
      add(p, 'formatting', 'Fix formatting issue', issue);
    }
  }

  // Contact
  if (!contact.details.email) add('critical', 'contact', 'Add email address', 'ATS requires a valid email to contact you.', null);
  if (!contact.details.phone) add('critical', 'contact', 'Add phone number', 'Include a phone number for recruiter outreach.', null);
  if (!contact.details.linkedin) add('important', 'contact', 'Add LinkedIn profile', 'Recruiters often verify candidates via LinkedIn.', null);
  if (!contact.details.location) add('nice', 'contact', 'Add your location', 'City and state help with location-based job matching.', null);

  // Structure
  for (const missing of structure.missing) {
    add('critical', 'structure', `Add ${missing} section`, `A "${missing}" section is expected by ATS parsers and recruiters.`, null);
  }

  // Achievements
  if (achievements.count === 0) {
    add('critical', 'achievements', 'Quantify your impact', 'Add numbers to your experience bullets (e.g., "Increased revenue by 35%").', 'experience',
      { before: 'Managed the sales team', after: 'Managed a 12-person sales team, driving $2.4M in quarterly revenue (+35% YoY)' });
  } else if (achievements.count < 3) {
    add('important', 'achievements', 'Add more metrics', `Only ${achievements.count} quantified achievement(s) found. Aim for 5+.`, 'experience');
  }

  // Action Verbs
  if (actionVerbs.weak.length > 0) {
    add('important', 'actionVerbs', 'Replace weak verbs', `Avoid: "${actionVerbs.weak.slice(0,3).join('", "')}". Use strong verbs like "Spearheaded", "Optimized", "Delivered".`, 'experience',
      { before: `Helped with the project`, after: `Spearheaded the project delivery` });
  }
  if (actionVerbs.coverage < 60) {
    add('important', 'actionVerbs', 'Start bullets with verbs', `Only ${actionVerbs.coverage}% of bullets start with action verbs. Aim for 80%+.`, 'experience');
  }

  // Length
  if (length.wordCount < 250) {
    add('critical', 'length', 'Resume is too short', 'Your resume has fewer than 250 words. Expand your experience and skills sections.', null);
  } else if (length.wordCount > 1200) {
    add('important', 'length', 'Consider shortening', 'Your resume exceeds 1,200 words. Trim older roles and focus on relevant experience.', null);
  }

  // Sort by priority
  const order = { critical: 0, important: 1, nice: 2 };
  suggestions.sort((a, b) => order[a.priority] - order[b.priority]);

  return suggestions.slice(0, 12);
}

// ── Domain Detection ─────────────────────────────────────────────

const DOMAIN_LABELS = {
  Software_Engineering: 'Software Engineering',
  Data_Science: 'Data Science & Analytics',
  Healthcare: 'Healthcare & Clinical',
  Finance: 'Finance & Accounting',
  Marketing: 'Marketing & Digital',
  Project_Management: 'Project Management',
  Human_Resources: 'Human Resources',
  Design_UX: 'Design & UX',
};

// Cache compiled regexes for word-boundary matching
const _regexCache = new Map();
function getWordBoundaryRegex(term) {
  if (_regexCache.has(term)) return _regexCache.get(term);
  // Escape regex special chars, then wrap with word boundaries
  const escaped = term.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const regex = new RegExp(`\\b${escaped}\\b`, 'i');
  _regexCache.set(term, regex);
  return regex;
}

function detectDomains(text) {
  const lower = text.toLowerCase();
  // Pre-tokenize text into a word set for fast exact matching of short terms
  const wordTokens = new Set(lower.split(/[\s,;|/()[\]{}<>:]+/).filter(w => w.length > 0));
  const results = [];

  for (const [domainKey, subcategories] of Object.entries(powerSkillsData)) {
    if (domainKey === 'Soft_Skills') continue;

    const domainSkills = deepFlattenSkills(subcategories).map(s => s.toLowerCase());
    const totalKeywords = domainSkills.length;
    if (totalKeywords === 0) continue;

    let matchCount = 0;
    const matchedTerms = [];

    for (const skill of domainSkills) {
      const hasSpace = skill.includes(' ');
      let isMatch = false;

      if (hasSpace) {
        // Multi-word phrase: use word-boundary regex to avoid partial overlap
        isMatch = getWordBoundaryRegex(skill).test(text);
      } else if (skill.length <= 2) {
        // Very short terms (R, C, AI, QI, HR, UX): require exact token match
        // This prevents "ai" matching inside "training", "r" inside "your", etc.
        isMatch = wordTokens.has(skill);
      } else if (skill.length <= 4) {
        // Short terms (SQL, CSS, HTML, Java, etc.): word boundary match
        isMatch = getWordBoundaryRegex(skill).test(text);
      } else {
        // Normal length terms: word boundary match
        isMatch = getWordBoundaryRegex(skill).test(text);
      }

      if (isMatch) {
        matchCount++;
        matchedTerms.push(skill);
      }
    }

    // Require minimum 3 keyword matches to register a domain
    if (matchCount < 3) continue;

    // Weighted confidence: raw match ratio boosted by absolute match count
    const rawRatio = matchCount / totalKeywords;
    const densityBonus = Math.min(0.20, matchCount * 0.006);
    const confidence = Math.min(99, Math.round((rawRatio + densityBonus) * 100));

    // Suppress noise: minimum 3% confidence
    if (confidence < 3) continue;

    results.push({
      domain: DOMAIN_LABELS[domainKey] || domainKey,
      key: domainKey,
      confidence,
      matchedCount: matchCount,
      totalKeywords,
      topMatches: [...new Set(matchedTerms)].slice(0, 8),
    });
  }

  results.sort((a, b) => b.confidence - a.confidence);
  return results.slice(0, 5);
}

// ── Main export ──────────────────────────────────────────────────

export function scoreResume(text, jobDescription = '') {
  if (!text || text.trim().length < 10) {
    return { total: 0, categories: {}, suggestions: [], keywords: { matched: [], missing: [] }, sections: [], domains: [] };
  }

  const keywords = scoreKeywords(text, jobDescription);
  const formatting = scoreFormatting(text);
  const contact = scoreContact(text);
  const structure = scoreStructure(text);
  const achievements = scoreAchievements(text);
  const actionVerbs = scoreActionVerbs(text);
  const length = scoreLength(text);

  const categories = { keywords, formatting, contact, structure, achievements, actionVerbs, length };
  const total = Math.min(100, Object.values(categories).reduce((s, c) => s + c.score, 0));
  const suggestions = generateSuggestions(categories);
  const sections = detectSections(text);
  const domains = detectDomains(text);

  return { total, categories, suggestions, keywords: { matched: keywords.matched, missing: keywords.missing }, sections, domains };
}

export { detectSections, detectDomains };

File: backend/ats_scorer.py
"""
ats_scorer.py — Backend ATS scoring engine.

Uses spaCy NLP, BERT embeddings, and rule-based analysis
to produce a comprehensive ATS score with domain detection.

Replaces frontend-only scoring with server-side intelligence.
"""

from __future__ import annotations

import re
import json
import hashlib
from collections import OrderedDict
from pathlib import Path
from typing import Any

import numpy as np

# ── Load domain skill database ───────────────────────────────────

_DATA_DIR = Path(__file__).resolve().parent.parent / "frontend" / "src" / "data"
_POWER_SKILLS_PATH = _DATA_DIR / "powerSkills.json"
_ACTION_VERBS_PATH = _DATA_DIR / "actionVerbs.json"
_SECTION_KW_PATH = _DATA_DIR / "sectionKeywords.json"

with _POWER_SKILLS_PATH.open() as f:
    POWER_SKILLS: dict = json.load(f)

with _ACTION_VERBS_PATH.open() as f:
    ACTION_VERBS_LIST: list[str] = json.load(f)

with _SECTION_KW_PATH.open() as f:
    SECTION_KEYWORDS: dict[str, list[str]] = json.load(f)


# ── Flatten helpers ──────────────────────────────────────────────

def _deep_flatten(obj: dict | list) -> list[str]:
    """Recursively flatten nested dicts/lists into a flat string list."""
    result: list[str] = []
    if isinstance(obj, dict):
        for val in obj.values():
            result.extend(_deep_flatten(val))
    elif isinstance(obj, list):
        for item in obj:
            if isinstance(item, str):
                result.append(item)
            else:
                result.extend(_deep_flatten(item))
    return result


ALL_SKILLS = list(set(s.lower() for s in _deep_flatten(POWER_SKILLS)))
ACTION_VERBS_SET = set(v.lower() for v in ACTION_VERBS_LIST)
WEAK_VERBS = {
    "helped", "worked", "did", "made", "was", "had", "got", "went",
    "used", "tried", "handled", "assisted", "participated",
    "responsible", "involved",
}

# ── Word-boundary matching ───────────────────────────────────────

_REGEX_CACHE: dict[str, re.Pattern] = {}


def _word_regex(term: str) -> re.Pattern:
    """Get or create a compiled word-boundary regex for a term."""
    if term not in _REGEX_CACHE:
        escaped = re.escape(term)
        _REGEX_CACHE[term] = re.compile(rf"\b{escaped}\b", re.IGNORECASE)
    return _REGEX_CACHE[term]


def _match_skill(skill: str, text_lower: str, word_tokens: set[str]) -> bool:
    """Match a skill term using context-appropriate strategy."""
    if " " in skill:
        return bool(_word_regex(skill).search(text_lower))
    elif len(skill) <= 2:
        return skill in word_tokens
    else:
        return bool(_word_regex(skill).search(text_lower))


def _tokenize(text: str) -> set[str]:
    return set(re.split(r"[\s,;|/()[\]{}<>:]+", text.lower()))


# ── Score: Keywords (30 pts) ─────────────────────────────────────

def score_keywords(
    text: str,
    spacy_skills: list[str],
    job_description: str = "",
) -> dict[str, Any]:
    """Score keyword coverage using NLP-extracted + regex-matched skills."""
    lower = text.lower()
    tokens = _tokenize(text)

    # 1. Regex-based matching against skill database
    regex_matched = [s for s in ALL_SKILLS if _match_skill(s, lower, tokens)]

    # 2. Merge with spaCy-extracted skills (higher quality)
    all_matched = list(set(regex_matched + [s.lower() for s in spacy_skills]))

    # 3. JD matching if provided
    jd_terms: list[str] = []
    if job_description:
        jd_lower = job_description.lower()
        jd_tokens = _tokenize(job_description)
        jd_terms = list(set(
            s for s in ALL_SKILLS if _match_skill(s, jd_lower, jd_tokens)
        ))

    expected = max(len(jd_terms) or 12, 12)
    source = jd_terms if jd_terms else ALL_SKILLS[:50]
    missing = [s for s in source if s not in all_matched][:15]

    pts = min(30, round((len(all_matched) / expected) * 30))
    return {
        "score": pts,
        "max": 30,
        "matched": all_matched[:50],
        "missing": missing,
        "matchPct": round((len(all_matched) / expected) * 100),
    }


# ── Score: Formatting (20 pts) ───────────────────────────────────

def score_formatting(text: str) -> dict[str, Any]:
    pts = 20
    issues: list[str] = []

    # Table-like layouts
    if re.search(r"[│┃┆┇┊┋|]{2,}", text) or re.search(r"\t.*\t.*\t", text):
        pts -= 5
        issues.append("Table-like layout detected — ATS often misreads tables")

    # Image references
    if re.search(r"\[image\]|\[graphic\]|\.png|\.jpg|\.svg", text, re.I):
        pts -= 5
        issues.append("Image/graphic references detected — ATS cannot parse images")

    # Overly long lines
    lines = text.split("\n")
    wide = [l for l in lines if len(l) > 120]
    if len(wide) > 3:
        pts -= min(4, len(wide))
        issues.append("Lines exceed 120 characters — may cause parsing issues")

    # Fancy bullets
    fancy = len(re.findall(r"[◆▶►★✦✧●○◉⬤⚫⬥◈▪▫]", text))
    if fancy > 0:
        pts -= min(3, fancy)
        issues.append("Fancy bullets detected — use standard dashes or dots")

    # Section headers
    found_headers = sum(
        1 for kws in SECTION_KEYWORDS.values()
        if any(kw in text.lower() for kw in kws)
    )
    if found_headers < 3:
        pts -= 5
        issues.append("Missing standard section headers")

    return {"score": max(0, pts), "max": 20, "issues": issues}


# ── Score: Contact (10 pts) ──────────────────────────────────────

def score_contact(text: str) -> dict[str, Any]:
    pts = 0
    details: dict[str, bool] = {}

    if re.search(r"[\w.-]+@[\w.-]+\.\w{2,}", text, re.I):
        pts += 3; details["email"] = True
    if re.search(r"(\+?\d[\d\s\-().]{7,}\d)", text):
        pts += 3; details["phone"] = True
    if re.search(r"linkedin\.com/in/[\w-]+", text, re.I):
        pts += 2; details["linkedin"] = True
    if re.search(r"\b[A-Z][a-z]+,?\s*[A-Z]{2}\b", text) or re.search(r"\b\d{5,6}\b", text):
        pts += 2; details["location"] = True
    if re.search(r"github\.com/[\w-]+", text, re.I):
        pts += 1

    return {"score": min(10, pts), "max": 10, "details": details}


# ── Score: Structure (15 pts) ────────────────────────────────────

def score_structure(text: str) -> dict[str, Any]:
    lower = text.lower()
    required = ["summary", "experience", "education", "skills"]
    found, missing = [], []

    for sec in required:
        kws = SECTION_KEYWORDS.get(sec, [])
        if any(kw in lower for kw in kws):
            found.append(sec)
        else:
            missing.append(sec)

    pts = round(len(found) * 3.75)
    return {"score": pts, "max": 15, "found": found, "missing": missing}


# ── Score: Achievements (15 pts) ─────────────────────────────────

_QUANT_RE = re.compile(
    r"\d+[%x×kmb$]|\$[\d,.]+|\d+\s*(?:users|clients|projects|customers|employees|members|teams|revenue|sales|leads)",
    re.I,
)


def score_achievements(text: str) -> dict[str, Any]:
    matches = _QUANT_RE.findall(text)
    count = len(matches)

    if count == 0:     pts = 0
    elif count <= 2:   pts = 6
    elif count <= 5:   pts = 10
    elif count <= 9:   pts = 13
    else:              pts = 15

    examples = [l.strip() for l in text.split("\n") if _QUANT_RE.search(l)][:5]
    return {"score": pts, "max": 15, "count": count, "examples": examples}


# ── Score: Action Verbs (5 pts) ──────────────────────────────────

def score_action_verbs(text: str) -> dict[str, Any]:
    lines = [
        l.strip() for l in text.split("\n")
        if re.match(r"^[-•–—*]", l.strip()) or (20 < len(l.strip()) < 200)
    ]
    if not lines:
        return {"score": 0, "max": 5, "coverage": 0, "weak": []}

    strong = 0
    weak_found: list[str] = []
    for line in lines:
        first = re.sub(r"^[-•–—*\d.)]+\s*", "", line).split()[0].lower() if line else ""
        if not first:
            continue
        if first in ACTION_VERBS_SET:
            strong += 1
        elif first in WEAK_VERBS:
            weak_found.append(first)

    coverage = round((strong / len(lines)) * 100) if lines else 0

    if coverage >= 80:   pts = 5
    elif coverage >= 60: pts = 3
    elif coverage >= 40: pts = 2
    else:                pts = 0

    return {"score": pts, "max": 5, "coverage": coverage, "weak": list(set(weak_found))}


# ── Score: Length (5 pts) ────────────────────────────────────────

def score_length(text: str) -> dict[str, Any]:
    wc = len(text.split())

    if 400 <= wc <= 700:     pts, msg = 5, "Optimal length for a 1-page resume"
    elif 701 <= wc <= 900:   pts, msg = 3, "Slightly dense — consider trimming"
    elif 901 <= wc <= 1200:  pts, msg = 5, "Good length for a 2-page resume"
    elif wc > 1200:          pts, msg = 2, "Too long — most ATS prefer 1-2 pages"
    elif wc < 250:           pts, msg = 2, "Too short — add more detail"
    else:                    pts, msg = 3, "Acceptable length"

    return {"score": pts, "max": 5, "wordCount": wc, "message": msg}


# ── Domain Detection (BERT-powered) ─────────────────────────────

DOMAIN_LABELS = {
    "Software_Engineering": "Software Engineering",
    "Data_Science": "Data Science & Analytics",
    "Healthcare": "Healthcare & Clinical",
    "Finance": "Finance & Accounting",
    "Marketing": "Marketing & Digital",
    "Project_Management": "Project Management",
    "Human_Resources": "Human Resources",
    "Design_UX": "Design & UX",
}


def detect_domains_nlp(
    text: str,
    spacy_skills: list[str],
    embedding: np.ndarray | None = None,
    sentence_model: Any = None,
) -> list[dict[str, Any]]:
    """
    Detect professional domains using:
    1. NLP skill matching (word-boundary + spaCy extracted)
    2. Optionally BERT embedding cosine to domain centroids
    """
    lower = text.lower()
    tokens = _tokenize(text)
    spacy_lower = set(s.lower() for s in spacy_skills)
    results = []

    for domain_key, subcategories in POWER_SKILLS.items():
        if domain_key == "Soft_Skills":
            continue

        domain_skills = [s.lower() for s in _deep_flatten(subcategories)]
        total = len(domain_skills)
        if total == 0:
            continue

        match_count = 0
        matched_terms: list[str] = []
        for skill in domain_skills:
            # Check via word-boundary regex or spaCy extraction
            if _match_skill(skill, lower, tokens) or skill in spacy_lower:
                match_count += 1
                matched_terms.append(skill)

        if match_count < 3:
            continue

        ratio = match_count / total
        bonus = min(0.20, match_count * 0.006)
        confidence = min(99, round((ratio + bonus) * 100))

        if confidence < 3:
            continue

        results.append({
            "domain": DOMAIN_LABELS.get(domain_key, domain_key),
            "key": domain_key,
            "confidence": confidence,
            "matchedCount": match_count,
            "totalKeywords": total,
            "topMatches": list(set(matched_terms))[:8],
        })

    results.sort(key=lambda d: d["confidence"], reverse=True)
    return results[:5]


# ── Suggestion Generator ─────────────────────────────────────────

def generate_suggestions(cats: dict[str, dict[str, Any]]) -> list[dict[str, Any]]:
    suggestions: list[dict[str, Any]] = []
    _id = 0

    def add(priority: str, category: str, title: str, detail: str,
            section_target: str | None = None, diff: dict | None = None):
        nonlocal _id
        suggestions.append({
            "id": f"s{_id}", "priority": priority, "category": category,
            "title": title, "detail": detail,
            "sectionTarget": section_target, "diff": diff,
        })
        _id += 1

    kw = cats["keywords"]
    fmt = cats["formatting"]
    contact = cats["contact"]
    struct = cats["structure"]
    ach = cats["achievements"]
    av = cats["actionVerbs"]
    ln = cats["length"]

    # Keywords
    if kw["score"] < 15:
        add("critical", "keywords", "Add more relevant skills",
            f"Only {len(kw['matched'])} skills detected. Add: {', '.join(kw['missing'][:5])}.",
            "skills")
    elif kw["score"] < 24:
        add("important", "keywords", "Expand your skillset",
            f"Missing key terms: {', '.join(kw['missing'][:4])}.", "skills")

    # Formatting
    for issue in fmt.get("issues", [])[:3]:
        p = "critical" if fmt["score"] < 10 else "important"
        add(p, "formatting", "Fix formatting issue", issue)

    # Contact
    d = contact.get("details", {})
    if not d.get("email"):
        add("critical", "contact", "Add email address",
            "ATS requires a valid email to contact you.")
    if not d.get("phone"):
        add("critical", "contact", "Add phone number",
            "Include a phone number for recruiter outreach.")
    if not d.get("linkedin"):
        add("important", "contact", "Add LinkedIn profile",
            "Recruiters often verify candidates via LinkedIn.")

    # Structure
    for m in struct.get("missing", []):
        add("critical", "structure", f"Add {m} section",
            f'A "{m}" section is expected by ATS parsers.')

    # Achievements
    if ach["count"] == 0:
        add("critical", "achievements", "Quantify your impact",
            'Add numbers: e.g., "Increased revenue by 35%".',
            "experience",
            {"before": "Managed the sales team",
             "after": "Managed a 12-person sales team, driving $2.4M in quarterly revenue (+35% YoY)"})
    elif ach["count"] < 3:
        add("important", "achievements", "Add more metrics",
            f"Only {ach['count']} quantified achievement(s). Aim for 5+.", "experience")

    # Action verbs
    if av.get("weak"):
        weak_str = '", "'.join(av["weak"][:3])
        add("important", "actionVerbs", "Replace weak verbs",
            f'Avoid: "{weak_str}". Use "Spearheaded", "Optimized", "Delivered".',
            "experience",
            {"before": "Helped with the project", "after": "Spearheaded the project delivery"})
    if av.get("coverage", 0) < 60:
        add("important", "actionVerbs", "Start bullets with verbs",
            f"Only {av['coverage']}% of bullets start with action verbs. Aim for 80%+.", "experience")

    # Length
    wc = ln.get("wordCount", 0)
    if wc < 250:
        add("critical", "length", "Resume is too short",
            "Fewer than 250 words. Expand experience and skills.")
    elif wc > 1200:
        add("important", "length", "Consider shortening",
            "Over 1,200 words. Trim older roles, focus on relevant experience.")

    order = {"critical": 0, "important": 1, "nice": 2}
    suggestions.sort(key=lambda s: order.get(s["priority"], 2))
    return suggestions[:12]


# ── Main scoring function ────────────────────────────────────────

# Simple LRU cache for repeated scoring of same text
_SCORE_CACHE: OrderedDict[str, dict] = OrderedDict()
_CACHE_MAX = 64


def score_resume(
    text: str,
    job_description: str = "",
    spacy_skills: list[str] | None = None,
    embedding: np.ndarray | None = None,
    sentence_model: Any = None,
) -> dict[str, Any]:
    """
    Full ATS scoring pipeline.
    
    Args:
        text: Raw resume text
        job_description: Optional JD for alignment scoring
        spacy_skills: Skills extracted via spaCy NLP (from skills.py)
        embedding: BERT embedding vector (if already computed)
        sentence_model: SentenceTransformer model (for domain BERT matching)
    
    Returns:
        ScoreResult with total, categories, domains, suggestions
    """
    if not text or len(text.strip()) < 10:
        return {
            "total": 0, "categories": {},
            "suggestions": [], "domains": [],
            "keywords": {"matched": [], "missing": []},
        }

    # Cache check
    text_hash = hashlib.md5((text + job_description).encode()).hexdigest()
    if text_hash in _SCORE_CACHE:
        return _SCORE_CACHE[text_hash]

    skills = spacy_skills or []

    # Run all scoring categories
    keywords = score_keywords(text, skills, job_description)
    formatting = score_formatting(text)
    contact = score_contact(text)
    structure = score_structure(text)
    achievements = score_achievements(text)
    action_verbs = score_action_verbs(text)
    length = score_length(text)

    categories = {
        "keywords": keywords,
        "formatting": formatting,
        "contact": contact,
        "structure": structure,
        "achievements": achievements,
        "actionVerbs": action_verbs,
        "length": length,
    }

    total = min(100, sum(c["score"] for c in categories.values()))
    suggestions = generate_suggestions(categories)
    domains = detect_domains_nlp(text, skills, embedding, sentence_model)

    result = {
        "total": total,
        "categories": categories,
        "suggestions": suggestions,
        "domains": domains,
        "keywords": {
            "matched": keywords["matched"],
            "missing": keywords["missing"],
        },
    }

    # Cache result
    _SCORE_CACHE[text_hash] = result
    if len(_SCORE_CACHE) > _CACHE_MAX:
        _SCORE_CACHE.popitem(last=False)

    return result
